File: jinja/go/genciconf.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"os"
	"strings"
	"text/template"

	"github.com/lithammer/dedent"
)

func getDSMList() []string {
        settingsGradle := "settings.gradle"
        marker := "include '"
        dsmList := []string{}

        file, err := os.Open(settingsGradle)
        if err != nil {
                fmt.Println(err)
                os.Exit(1)
        }
        defer file.Close()

        scanner := bufio.NewScanner(file)
        for scanner.Scan() {
                lineText := scanner.Text()
                lineSlice := strings.Split(lineText, "\n")
                for _, line := range lineSlice {
                        if strings.Contains(line, marker) {
                                dsmList = append(dsmList, line[9:len(line)-1])
                        }
                }
        }

        return dsmList
}


type DSM struct {
        Name string
}

func getTriggerSnippet(dsm string) string {
        dsmName := DSM{dsm}

        // TODO: (tranh) avoid empty first line
        triggerTemplate := `
                Build {{ .Name }}:
                  stage: Triggers
                  trigger:
                    include: {{ .Name }}/.gitlab-ci.yml
                    strategy: depend
                  rules:
                    - changes:
                        - **/*`

        // Removes common leading whitespace from multiline strings
        t, err := template.New("dsmTemplate").Parse(dedent.Dedent(triggerTemplate))
        if err != nil {
                fmt.Println(err)
                os.Exit(1)
        }

        // Create a buffer to write our result to.
        var templateOutputBuf bytes.Buffer
        err = t.Execute(&templateOutputBuf, dsmName)

        if err != nil {
                fmt.Println(err)
                os.Exit(1)
        }

        triggerText := templateOutputBuf.String()

        return triggerText
}

func genTriggerYaml() {
        triggerYaml := "triggers.yml"
        dsmList := getDSMList()

        for _, dsm := range dsmList {
                // Appending text to file
                f, err := os.OpenFile(triggerYaml, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
                if err != nil {
                        fmt.Println(err)
                        os.Exit(1)
                }
                defer f.Close()

                if _, err := f.WriteString(getTriggerSnippet(dsm) + "\n"); err != nil {
                        fmt.Println(err)
                        os.Exit(1)
                }
        }
}


func main() {
        genTriggerYaml()
}

File: jinja/cpp/genciconf.cpp
#include <fstream>
#include <inja/inja.hpp>
#include <iostream>
#include <nlohmann/json.hpp>
#include <stdlib.h>
#include <string>
#include <vector>

std::vector<std::string> getDSMList(void) {
  std::string settingsGradleFile = "settings.gradle";
  std::string marker = "include '";
  std::vector<std::string> dsmList = {};

  std::ifstream file(settingsGradleFile);
  if (!file.is_open()) {
    std::cerr << "No such file: " << settingsGradleFile << std::endl;
    exit(255);
  }
  std::string line;
  while (std::getline(file, line)) {
    if (line.find(marker) != std::string::npos) {
      // TODO: I don't know why line.length() - 10 works here :-/
      dsmList.push_back(line.substr(9, line.length() - 10));
    }
  }
  file.close();

  return dsmList;
}

std::string getTriggerSnippet(std::string dsm) {
  nlohmann::json j;

  j["dsm"] = dsm;

  std::string templateStr = "Build {{ dsm }}:\n"
                            "  stage: Triggers\n"
                            "  trigger:\n"
                            "    include: {{ dsm }}/.gitlab-ci.yml\n"
                            "    strategy: depend\n"
                            "  rules:\n"
                            "    - changes:\n"
                            "        - \"**/*\"\n";

  std::string renderedText = inja::render(templateStr, j);

  return renderedText;
}

void genTriggerYaml(void) {
  std::string triggerYaml = "triggers.yml";
  std::vector<std::string> dsmList = getDSMList();

  std::fstream file;
  // Opens file for appending
  file.open(triggerYaml, std::ios_base::app);
  if (!file.is_open()) {
    std::cerr << "Unable to open: " << triggerYaml << std::endl;
    exit(255);
  }

  for (int i = 0; i < dsmList.size(); i++) {
    file << getTriggerSnippet(dsmList[i]) << std::endl;
  }
  file.close();
}

int main(int argc, char **argv) {
  genTriggerYaml();

  return 0;
}
